### PR TITLE
RATEST:312: Password Format verification and combining to RATEST 157

### PR DESCRIPTION
**Pull request by mherman22**
**Ticket-ID**: https://issues.openmrs.org/browse/RATEST-312

Workflow for the following:

- Password Requires Upper And Lower Case Test
- Password Requires Non Digit Test
- Password Requires Digit Test
- Password Mininum Length Test

### CODE REVIEW

**Comment by mherman22** [comment]
cc: @sherrif10 @kdaud

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java [hunk 1]
```diff
@@ -32,6 +32,7 @@
 	private static final String CLERK = RandomStringUtils.randomAlphabetic(8);
 	private static final String NURSE = RandomStringUtils.randomAlphabetic(8);
 	private static final String DOCTOR = RandomStringUtils.randomAlphabetic(8);
+	private static final String LAB_ASSISTANT = RandomStringUtils.randomAlphanumeric(8);
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java:35]: any reason for adding another user account ?

Comment (mherman22) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java:35]: lemi change that and use the available accounts

Comment (mherman22) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java]: resolved!!

**Comment by kdaud** [comment]
Here is my thought! 

We should have two test cases; when a user creates password that meets the password criteria then the system should save the password and the opposite should be true. This mean we can have two user stories for the test.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/06-user/users.feature [hunk 2]
```diff
@@ -31,32 +31,16 @@ Feature: User Account Management
 
   @selenium
   @userAccount
-  Scenario: Password Requires Upper and Lower case
+  Scenario: Passwords that dont meet the password criteria
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/06-user/users.feature:34]: isn't this suppossed to be `Password that doesn't meet ...`?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/06-user/users.feature [hunk 3]
```diff
@@ -31,32 +31,16 @@ Feature: User Account Management
...
     And a user enters person details in the user account form
     And a user enters user account details in the user account form
-    And a user sets the passwords with only upper case letters
+    And a user sets the passwords that dont meet the password criteria
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/06-user/users.feature:37]: isn't this suppossed to be` And a user enters the password that doesn't ... `?

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/06-user/users.feature [hunk 4]
```diff
@@ -31,32 +31,16 @@ Feature: User Account Management
...
     Then the system throws a validation error
 
   @selenium
   @userAccount
-  Scenario: Password Requires Non-digits
+  Scenario: Passwords that meet the password criteria
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/06-user/users.feature:42]: Is this supposed to be plural or singular?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java [hunk 5]
```diff
@@ -146,4 +146,38 @@ public void clickOnSaveUserAccountButton() {
...
+	}
+
+	@And("a user enters user account details in the user account form")
+	public void enterUserAccountDetails() {
+		userAccountPage.clickOnAddUserAccount();
+		userAccountPage.setUsername(CLERK);
+		userAccountPage.setUserPrivilegeLevel("Full");
+	}
+
+	@And("a user sets the password that meets the password criteria")
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java:163]: Isn't this supposed to be `And a user enters password that .. `?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java [hunk 6]
```diff
@@ -146,4 +146,38 @@ public void clickOnSaveUserAccountButton() {
...
+	public void userSetsPasswordThatMeetsThePasswordCriteria() {
+		userAccountPage.setUserPassword("Dataclerk!23", "Dataclerk!23");
+	}
+	
+	@And("a user sets the password that doesnt meet the password criteria")
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java:168]: even here

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java [hunk 7]
```diff
@@ -146,4 +146,38 @@ public void clickOnSaveUserAccountButton() {
...
+		userAccountPage.setUsername(CLERK);
+		userAccountPage.setUserPrivilegeLevel("Full");
+	}
+
+	@And("a user enters password that meets the password criteria")
+	public void userEnterPasswordThatMeetsThePasswordCriteria() {
+		userAccountPage.setUserPassword("Dataclerk!23", "Dataclerk!23");
+	}
+	
+	@And("a user enters password that doesnt meet the password criteria")
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java:168]: What  do you think of `doesnt` vs `doesn't`

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java [hunk 8]
```diff
@@ -146,4 +146,38 @@ public void clickOnSaveUserAccountButton() {
...
+	public void userEntersPasswordThatDoesntMeetThePasswordCriteria() {
+		userAccountPage.setUserPassword("DATACLERK!23", "DATACLERK!23");
+	}
+	
+	@Then("the system saves the password")
+	public void systemSavesThePassword() {
+		assertTrue(textExists("Account Saved Successfully"));
+	}
+
+	 @Then("the system throws a validation error")
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java:178]:
```suggestion
	 @Then("the system throws a validation error message")
```

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java [hunk 9]
```diff
@@ -146,4 +146,38 @@ public void clickOnSaveUserAccountButton() {
...
+	 public void systemThrowsAvalidationError() {
+		 List<String> validationErrors = userAccountPage.getValidationErrors();
+	 	assertTrue(userAccountPage.isDataCorrect(validationErrors));
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java:181]: Fix the indention here!

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java [hunk 10]
```diff
@@ -146,4 +146,38 @@ public void clickOnSaveUserAccountButton() {
...
+	@And("a user enters person details in the user account form") 
+	public void enterPersonalDetails() {
+		userAccountPage.enterPersonalDetails("Clerk", "Data");
+		userAccountPage.selectGender();
+	}
+
+	@And("a user enters user account details in the user account form")
+	public void enterUserAccountDetails() {
+		userAccountPage.clickOnAddUserAccount();
+		userAccountPage.setUsername(CLERK);
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java:159]: We can use a random user name.

Comment (mherman22) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/UserAccountSteps.java:159]: ts randomly generated already, take a look at `private static final String CLERK = RandomStringUtils.randomAlphabetic(8);`

File: .github/workflows/platform-install-advanced.yml [hunk 11]
```diff
@@ -40,9 +40,9 @@ jobs:
         run: while ! nc -z localhost 8080 </dev/null; do sleep 1; done
       - name: Run qaframework on selenium
         run: |
-          echo "cucumber.publish.enabled=true" > qaframework-bdd-tests/src/test/resources/cucumber.properties
-          mvn clean install -DskipTests=true
-          cp -f qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test-local.properties qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test.properties
           cd qaframework-bdd-tests
+          echo "cucumber.publish.enabled=true" > src/test/resources/cucumber.properties
+          mvn clean install -DskipTests=true
+          cp -f src/test/resources/org/openmrs/uitestframework/test-local.properties src/test/resources/org/openmrs/uitestframework/test.properties
```
Comment (kdaud) [inline .github/workflows/platform-install-advanced.yml:46]: Are all these changes part of the ticket requirements?

**Pull request closed without merging**